File: packages/api/src/crowdflow_api/spectator.py
"""Build the deliberately small spectator feed from served engine conclusions.

This adapter translates a live ``TickEnvelope`` plus a route request into the
Pydantic ``SpectatorView`` contract.  It does not classify density, choose an
intervention or bypass safety.  Bands come from ``ZoneState``; routes come from
``VenueGraph``; reroutes are exposed only when the exact command was dispatched
through ``SafetyEngine.review()`` by the control loop.

The phone receives one decision, not the operator state.  Unknown legs stay
unknown, crossing availability comes from the pack, and route duration is priced
once here rather than reconstructed in TypeScript.
"""

from __future__ import annotations

from crowdflow_contracts import (
    AheadView,
    CrossingClosed,
    CrossingNotice,
    CrossingOpen,
    LinkStatus,
    OfflineView,
    RerouteOffer,
    Route,
    SpectatorView,
    Step,
    WalkView,
    WayAhead,
)
from crowdflow_core.routing import VenueGraph

from .session import ScenarioSession
from .wire import TickEnvelope


class SpectatorFeedUnavailable(LookupError):
    """The requested route cannot honestly be rendered from current evidence."""


def _label(graph: VenueGraph, zone_id: str) -> str:
    zone = graph.pack.zones.get(zone_id)
    return zone.name if zone and zone.name else zone_id


def _edge_between(graph: VenueGraph, source: str, destination: str) -> str | None:
    return next((eid for nxt, eid in graph.neighbours(source) if nxt == destination), None)


def _crossing_for_edge(
    graph: VenueGraph,
    edge_id: str,
    *,
    session_state: str | None,
) -> CrossingNotice | None:
    crossing = next(
        (item for item in graph.pack.crossings.values() if item.edge_id == edge_id),
        None,
    )
    if crossing is None:
        return None
    is_open = crossing.availability.is_open_during(session_state)
    state = CrossingOpen(open=True, closes_at=None) if is_open else CrossingClosed(
        open=False, opens_at=None
    )
    return CrossingNotice(name=crossing.id.replace("_", " "), state=state)


def build_route(
    graph: VenueGraph,
    envelope: TickEnvelope,
    *,
    origin: str,
    destination: str,
    avoid: set[str] | None = None,
    prefer: set[str] | None = None,
    route_id: str = "current-route",
) -> Route:
    """Price one route and attach already-computed crowd conclusions to its legs."""
    result = graph.route(origin, destination, avoid=avoid, prefer=prefer)
    if not result.found:
        raise SpectatorFeedUnavailable(result.rejected_reason or "no route")

    steps: list[Step] = []
    session_state = envelope.state.session_id
    for index, (source, target) in enumerate(zip(result.path, result.path[1:], strict=False)):
        edge_id = _edge_between(graph, source, target)
        if edge_id is None:
            raise SpectatorFeedUnavailable(f"route edge missing: {source} -> {target}")
        edge = graph.pack.edges[edge_id]
        state = envelope.state.zones.get(target) or envelope.state.zones.get(source)
        way = WayAhead(state.band.value) if state is not None else WayAhead.UNKNOWN
        measured_speed = state.mean_speed_ms if state is not None else None
        speed = measured_speed or (
            edge.free_speed_ms.value if edge.free_speed_ms is not None else None
        )
        # ``result.eta_s`` already contains the standards free-speed fallback.
        # Allocate that total by edge length when this leg has no observed speed,
        # rather than introducing a second fallback constant here.
        if speed is not None and speed > 0:
            walk_s = edge.length_m / speed
        elif result.distance_m > 0:
            walk_s = result.eta_s * edge.length_m / result.distance_m
        else:
            walk_s = 0.0
        steps.append(
            Step(
                id=f"{route_id}-leg-{index}",
                to=_label(graph, target),
                walk_s=round(walk_s, 1),
                way_ahead=way,
                crossing=_crossing_for_edge(
                    graph, edge_id, session_state=session_state
                ),
            )
        )

    return Route.model_validate(
        {
            "id": route_id,
            "from": _label(graph, origin),
            "to": _label(graph, destination),
            "steps": steps,
            "total_walk_s": round(result.eta_s, 1),
        }
    )


def build_spectator_view(
    session: ScenarioSession,
    *,
    origin: str,
    destination: str,
    online: bool = True,
    mesh_peers: int = 0,
) -> SpectatorView:
    """Current walk/offline/ahead view for one route request.

    ``online`` and ``mesh_peers`` are transport observations supplied by the
    requesting device. They never alter a band or route; they only select the
    honest freshness treatment on the phone.
    """
    envelope = session.last_envelope
    if envelope is None:
        raise SpectatorFeedUnavailable("no crowd tick is available yet")
    if mesh_peers < 0:
        raise ValueError("mesh_peers must be non-negative")

    graph = session.circuit.graph
    route = build_route(
        graph,
        envelope,
        origin=origin,
        destination=destination,
    )
    link = LinkStatus(
        online=online,
        mesh_peers=mesh_peers,
        updated_at=envelope.state.timestamp,
    )

    if not online:
        return SpectatorView(
            root=OfflineView(kind="offline", now=envelope.time_s, link=link, route=route)
        )

    command = envelope.command
    verdict = envelope.verdict
    path = graph.route(origin, destination).path
    if (
        envelope.dispatched
        and command is not None
        and verdict is not None
        and command.is_valid_at(envelope.time_s)
        and command.source_zone in path
    ):
        instead = build_route(
            graph,
            envelope,
            origin=origin,
            destination=destination,
            avoid=set(command.avoid),
            prefer=set(command.prefer),
            route_id=f"{route.id}-reroute",
        )
        step_index = path.index(command.source_zone)
        if not route.steps:
            raise SpectatorFeedUnavailable("an affected route has no walkable leg")
        step_id = route.steps[min(step_index, len(route.steps) - 1)].id
        return SpectatorView(
            root=AheadView(
                kind="ahead",
                now=envelope.time_s,
                link=link,
                route=route,
                step_id=step_id,
                offer=RerouteOffer(command=command, verdict=verdict, instead=instead),
            )
        )

    return SpectatorView(
        root=WalkView(kind="walk", now=envelope.time_s, link=link, route=route)
    )

File: packages/api/src/spectator.ts
import type { AheadView, CircuitPack, CrossingNotice, LinkStatus, RerouteCommand, Route, SafetyVerdict, SpectatorView, Step, VenueState, WayAhead } from '@crowdflow/contracts';
import { isOpenDuring, validCommandAt } from '@crowdflow/contracts';
import { VenueGraph } from '@crowdflow/core';
import type { ScenarioSession } from './session.js';
import type { TickEnvelope } from './wire.js';

export class SpectatorFeedUnavailable extends Error {}
export const SPECTATOR_OFFER_RETENTION_S = 300;
export interface ActiveOffer { command: RerouteCommand; verdict: SafetyVerdict; dispatched_at: number }

export class SpectatorFeed {
  private offers = new Map<string, ActiveOffer>();
  constructor(readonly session: ScenarioSession) {}
  observe(envelope: TickEnvelope): void {
    if (envelope.dispatched && envelope.command && envelope.verdict?.dispatchable) this.offers.set(envelope.command.command_id, { command: envelope.command, verdict: envelope.verdict, dispatched_at: envelope.time_s });
    for (const [id, offer] of this.offers) if (!validCommandAt(offer.command, envelope.time_s) || envelope.time_s - offer.dispatched_at > SPECTATOR_OFFER_RETENTION_S) this.offers.delete(id);
  }
  view(request: SpectatorRequest): SpectatorView {
    const envelope = this.session.lastEnvelope; if (!envelope) throw new SpectatorFeedUnavailable('no crowd tick is available yet'); this.observe(envelope);
    return buildSpectatorView(this.session.circuit.graph, envelope, request, [...this.offers.values()]);
  }
}

export interface SpectatorRequest { origin: string; destination: string; online?: boolean; mesh_peers?: number; now_unix_s?: number }
export function buildSpectatorView(graph: VenueGraph, envelope: TickEnvelope, request: SpectatorRequest, offers: ActiveOffer[] = []): SpectatorView {
  const online = request.online ?? true; const peers = request.mesh_peers ?? 0; if (peers < 0) throw new Error('mesh_peers must be non-negative');
  const route = buildRoute(graph, envelope.state, request.origin, request.destination); const now = request.now_unix_s ?? Date.now() / 1000;
  // Simulation timestamps are relative. Freshness is translated once at the API boundary.
  const ageS = Math.max(0, envelope.time_s - envelope.state.timestamp); const link: LinkStatus = { online, mesh_peers: peers, updated_at: now - ageS };
  if (!online) return { kind: 'offline', now, link, route };
  const path = graph.route(request.origin, request.destination).path;
  const active = offers.filter((offer) => offer.verdict.dispatchable && validCommandAt(offer.command, envelope.time_s) && path.includes(offer.command.source_zone)).sort((a, b) => b.dispatched_at - a.dispatched_at || a.command.command_id.localeCompare(b.command.command_id))[0];
  if (!active) return { kind: 'walk', now, link, route };
  const instead = buildRoute(graph, envelope.state, request.origin, request.destination, new Set(active.command.avoid ?? []), new Set(active.command.prefer ?? []), `${route.id}-reroute`);
  const step = Math.min(path.indexOf(active.command.source_zone), route.steps.length - 1); if (step < 0) throw new SpectatorFeedUnavailable('an affected route has no walkable leg');
  return { kind: 'ahead', now, link, route, step_id: route.steps[step]!.id, offer: { command: active.command, verdict: active.verdict, instead } } satisfies AheadView;
}

export function buildRoute(graph: VenueGraph, state: VenueState, origin: string, destination: string, avoid = new Set<string>(), prefer = new Set<string>(), routeId = 'current-route'): Route {
  const result = graph.route(origin, destination, undefined, avoid, prefer); const steps: Step[] = [];
  for (let index = 0; index < result.path.length - 1; index += 1) {
    const source = result.path[index]!; const target = result.path[index + 1]!; const edge = selectEdge(graph.pack, source, target, state.session_id ?? null, avoid);
    if (!edge) throw new SpectatorFeedUnavailable(`route edge missing: ${source} -> ${target}`);
    const zone = state.zones?.[target] ?? state.zones?.[source]; const way = (zone?.band ?? 'unknown') as WayAhead; const speed = zone?.mean_speed_ms && zone.mean_speed_ms > 0 ? zone.mean_speed_ms : edge.free_speed_ms?.value;
    const walk = speed && speed > 0 ? edge.length_m / speed : result.distance_m > 0 ? result.eta_s * edge.length_m / result.distance_m : 0;
    steps.push({ id: `${routeId}-leg-${index}`, to: label(graph.pack, target), walk_s: Number(walk.toFixed(1)), way_ahead: way, crossing: crossing(graph.pack, edge.id, state.session_id ?? null) });
  }
  return { id: routeId, from: label(graph.pack, origin), to: label(graph.pack, destination), steps, total_walk_s: Number(result.eta_s.toFixed(1)) };
}

function selectEdge(pack: CircuitPack, source: string, destination: string, session: string | null, avoid: Set<string>) {
  return Object.values(pack.edges ?? {}).filter((edge) => !avoid.has(edge.id) && ((edge.source === source && edge.destination === destination) || ((edge.bidirectional ?? true) && edge.source === destination && edge.destination === source))).filter((edge) => { const item = Object.values(pack.crossings ?? {}).find((value) => value.edge_id === edge.id); return !item?.availability || isOpenDuring(item.availability, session); }).sort((a, b) => a.length_m - b.length_m || a.id.localeCompare(b.id))[0];
}
function crossing(pack: CircuitPack, edgeId: string, session: string | null): CrossingNotice | null {
  const item = Object.values(pack.crossings ?? {}).find((value) => value.edge_id === edgeId); if (!item) return null; const open = !item.availability || isOpenDuring(item.availability, session); return { name: item.id.replaceAll('_', ' '), state: open ? { open: true, closes_at: null } : { open: false, opens_at: null } };
}
function label(pack: CircuitPack, id: string): string { return pack.zones?.[id]?.name ?? id; }
